File: solutions/JavaScript/invert_binary_tree_bfs.js
function TreeNode(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
}

/**
 * BFS Iterative approach to invert a binary tree.
 * Uses a queue to visit nodes level by level, swapping children.
 *
 * Time Complexity: O(n) - visit each node once
 * Space Complexity: O(w) - w = max width of tree (nodes at widest level)
 *
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function(root) {
    if (root === null) {
        return root;
    }

    const queue = [root];
    let front = 0;

    while (front < queue.length) {
        const node = queue[front++];

        // Swap left and right children
        [node.left, node.right] = [node.right, node.left];

        // Add children to queue for processing
        if (node.left !== null) {
            queue.push(node.left);
        }
        if (node.right !== null) {
            queue.push(node.right);
        }
    }

    return root;
};

// Test case
if (require.main === module) {
    // Create tree:     1
    //                /   \
    //               2     3
    const root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);

    invertTree(root);

    // Expected:        1
    //                /   \
    //               3     2
    console.log(`Root: ${root.val}`);  // 1
    console.log(`Left: ${root.left.val}, Right: ${root.right.val}`);  // 3, 2
}

module.exports = invertTree;

File: solutions/C++/invert_binary_tree_bfs.cpp
#include <iostream>
#include <queue>

using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    /**
     * BFS Iterative approach to invert a binary tree.
     * Uses a queue to visit nodes level by level, swapping children.
     *
     * Time Complexity: O(n) - visit each node once
     * Space Complexity: O(w) - w = max width of tree (nodes at widest level)
     */
    TreeNode* invertTree(TreeNode* root) {
        if (root == nullptr) {
            return root;
        }

        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();

            // Swap left and right children
            swap(node->left, node->right);

            // Add children to queue for processing
            if (node->left != nullptr) {
                q.push(node->left);
            }
            if (node->right != nullptr) {
                q.push(node->right);
            }
        }

        return root;
    }
};

int main() {
    // Create tree:     1
    //                /   \
    //               2     3
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);

    Solution sol;
    sol.invertTree(root);

    // Expected:        1
    //                /   \
    //               3     2
    cout << "Root: " << root->val << endl;  // 1
    cout << "Left: " << root->left->val << ", Right: " << root->right->val << endl;  // 3, 2

    return 0;
}
